src/partial_quicksync.rs

```diff
@@ -0,0 +1,265 @@
+    .window_log_max(31)
+    .context("Failed to set window log max")?;
+
+  io::copy(&mut decoder, &mut writer).context("Failed to decompress")?;
+
+  Ok(())
+}
+
+fn execute_restore(target_conn: &Connection, restore_string: &str) -> Result<()> {
+  target_conn
+    .execute_batch(restore_string)
+    .context("Failed to execute restore")
+}
+
+pub fn partial_restore(layer_from: i64, target_db_path: &str, jump_back: usize) -> Result<()> {
+  let client = Client::new();
+  let base_url = get_base_url();
+  let user_version = get_user_version(target_db_path)?;
+  let remote_metadata = client
+    .get(format!("{}/{}/metadata.csv", base_url, user_version))
+    .send()?
+    .text()?;
+  let restore_string = client
+    .get(format!("{}/{}/restore.sql", base_url, user_version))
+    .send()?
+    .text()?;
+  let layer_from = if layer_from == 0 {
+    get_latest_from_db(target_db_path)?
+  } else {
+    layer_from
+  };
+  let start_points = find_start_points(layer_from, &remote_metadata, jump_back);
+  if start_points.is_empty() {
+    anyhow::bail!("No suitable restore point found");
+  }
+  println!("Found {} potential restore points", start_points.len());
+
+  for point in start_points {
+    let conn = Connection::open(target_db_path)?;
+    let previous_hash = get_previous_hash(point.from, &conn)?;
+
+    if previous_hash == &point.hash[..4] {
+      let source_db_path_zst = "backup_source.db.zst";
+      let source_db_path = "backup_source.db";
+
+      if download_file(
+        &client,
+        user_version,
+        point.from,
+        point.to,
+        &point.hash,
+        source_db_path_zst,
+      )
+      .is_err()
+      {
+        download_file(
+          &client,
+          user_version,
+          point.from,
+          point.to,
+          &point.hash,
+          source_db_path,
+        )?;
+      } else {
+        decompress_file(source_db_path_zst, source_db_path)?;
+        fs::remove_file(source_db_path_zst)?;
+      }
+
+      println!("Restoring from {} to {}...", point.from, point.to);
+      let start = Instant::now();
+      execute_restore(&conn, &restore_string)?;
+      fs::remove_file(source_db_path)?;
+      let duration = start.elapsed();
+      println!("Restored {} to {} in {:?}", point.from, point.to, duration);
+    }
+  }
+  Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use anyhow::Result;
+  use rusqlite::Connection;
+  use tempfile::tempdir;
+
+  fn create_test_db() -> Result<(tempfile::TempDir, String)> {
+    let dir = tempdir()?;
+    let db_path = dir.path().join("test.db");
+    let conn = Connection::open(&db_path)?;
+    conn.execute(
+      "CREATE TABLE layers (id INTEGER, applied_block INTEGER, aggregated_hash BLOB)",
+      [],
+    )?;
+    conn.execute(
+      "INSERT INTO layers (id, applied_block, aggregated_hash) VALUES (?, ?, ?), (?, ?, ?)",
+      rusqlite::params![1, 100, 0xAAAA, 2, 200, 0xBBBB],
+    )?;
+    Ok((dir, db_path.to_str().unwrap().to_string()))
+  }
```

Comment:
How about changing the `get_user_version()` and `get_latest_from_db()` to take ` Connection` and returning a `Connection` to an in-memory database here instead? Benefits:
- the mentioned functions indicate they operate on DB by taking a connection to the DB,
- DB connection can be reused,
- less boilerplate, simpler code

Btw, IMHO it's fine to unwrap() in a test helper.

```suggestion
  fn create_test_db() -> Connection {
    let conn = Connection::open_in_memory().unwrap();
    conn.execute(
      "CREATE TABLE layers (id INTEGER, applied_block INTEGER, aggregated_hash BLOB)",
      [],
    ).unwrap();
    conn.execute(
      "INSERT INTO layers (id, applied_block, aggregated_hash) VALUES (?, ?, ?), (?, ?, ?)",
      rusqlite::params![1, 100, 0xAAAA, 2, 200, 0xBBBB],
    ).unwrap();
  conn
  }
```

With it, the tests become simpler:

```rust
  #[test]
  fn test_get_latest_from_db() {
    let conn = create_test_db();
    let result = get_latest_from_db(&conn).unwrap();
    assert_eq!(result, 2);
  }

  #[test]
  fn test_get_user_version()  {
    let conn = create_test_db();
    conn.execute("PRAGMA user_version = 42", []).unwrap();
    let result = get_user_version(&conn).unwrap();
    assert_eq!(result, 42);
  }
```